File: 636.exclusive-time-of-functions.py
#
# @lc app=leetcode id=636 lang=python3
#
# [636] Exclusive Time of Functions
#

# @lc code=start
from typing import List
class Solution:
    def exclusiveTime(self, n: int, logs: List[str]) -> List[int]:
        #it really looks like a peeling onion problem
        def parse(log: str):
            idx, state, timestamp = log.split(':')
            return int(idx), state, int(timestamp)
        
        stack, ans, total = [], [0] * n, 0
        for log in logs:
            idx, state, timestamp = parse(log)
            if state == 'start':
                #push the tuple to the stack
                stack.append((idx, timestamp, total))
            else:
                idx_, timestamp_, total_ = stack.pop()
                diff = timestamp - timestamp_ + 1 - total + total_
                ans[idx] += diff
                total += diff
        return ans
# @lc code=end

File: 636.exclusive-time-of-functions.java
/*
 * @lc app=leetcode id=636 lang=java
 *
 * [636] Exclusive Time of Functions
 */
import java.util.*;
// @lc code=start
class Solution {
    public int[] exclusiveTime(int n, List<String> logs) {
        Deque<Integer> stack = new ArrayDeque<>();//store the idx of the function
        int[] ans = new int[n];
        int cur = -1; //this remembers the last timestamp
        for(String log : logs){
            String[] ss = log.split(":");
            int idx = Integer.parseInt(ss[0]), ts = Integer.parseInt(ss[2]);//ss[1] indicates start or end
            if(ss[1].equals("start")){
                if(!stack.isEmpty()) ans[stack.peek()] += ts - cur;
                stack.push(idx);
                cur = ts;
            }else{
                int index = stack.pop();
                ans[index] += ts - cur + 1;
                cur = ts + 1;
            }
        }
        return ans;
    }
}
// @lc code=end

